File: Кучи.py
#Бинарная куча
import heapq

# Min-heap пример
heap = []
heapq.heappush(heap, 5)
heapq.heappush(heap, 2)
heapq.heappush(heap, 8)
print(heapq.heappop(heap))  # 2

# Max-heap (используем отрицательные значения)
max_heap = []
heapq.heappush(max_heap, -5)
heapq.heappush(max_heap, -2)
heapq.heappush(max_heap, -8)
print(-heapq.heappop(max_heap))  # 8


#Биноминальная куча
class BinomialHeap:
    class Node:
        def init(self, key):
            self.key = key
            self.degree = 0
            self.parent = None
            self.child = None
            self.sibling = None
    
    def init(self):
        self.head = None
    
    def merge(self, h1, h2):
        # Реализация слияния биномиальных куч
        pass
    
    def insert(self, key):
        new_heap = BinomialHeap()
        new_heap.head = self.Node(key)
        self.head = self.merge(self.head, new_heap.head)
    
    def extract_min(self):
        # Реализация извлечения минимума
        pass


#Хеш таблица
# Встроенная хеш-таблица (dict)
hash_table = {}
hash_table["apple"] = 1
hash_table["banana"] = 2
hash_table["orange"] = 3

print(hash_table["apple"])  # 1
print("banana" in hash_table)  # True

# Удаление
del hash_table["orange"]




#Куча Фибоначи
class FibonacciHeap:
    class Node:
        def init(self, key):
            self.key = key
            self.degree = 0
            self.marked = False
            self.parent = None
            self.child = None
            self.left = self
            self.right = self
    
    def init(self):
        self.min_node = None
        self.count = 0
    
    def insert(self, key):
        new_node = self.Node(key)
        if self.min_node is None:
            self.min_node = new_node
        else:
            self._add_to_root_list(new_node)
            if key < self.min_node.key:
                self.min_node = new_node
        self.count += 1
        return new_node
    
    def _add_to_root_list(self, node):
        # Добавление узла в корневой список
        pass

File: Кучи.java
#Бинарная куча
import java.util.PriorityQueue;

public class BinaryHeapExample {
    public static void main(String[] args) {
        // Min-heap
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.add(5);
        minHeap.add(2);
        minHeap.add(8);
        System.out.println(minHeap.poll()); // 2

        // Max-heap
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        maxHeap.add(5);
        maxHeap.add(2);
        maxHeap.add(8);
        System.out.println(maxHeap.poll()); // 8
    }
}



#Биноминальная куча
import java.util.*;

public class BinomialHeap {
    class Node {
        int key;
        int degree;
        Node parent, child, sibling;
        
        Node(int key) {
            this.key = key;
            this.degree = 0;
        }
    }
    
    private Node head;
    
    public void insert(int key) {
        BinomialHeap newHeap = new BinomialHeap();
        newHeap.head = new Node(key);
        head = merge(head, newHeap.head);
    }
    
    private Node merge(Node h1, Node h2) {
        // Реализация слияния
        return null;
    }
}

#Хеш таблица
import java.util.*;

public class HashTableExample {
    public static void main(String[] args) {
        HashMap<String, Integer> hashTable = new HashMap<>();
        hashTable.put("apple", 1);
        hashTable.put("banana", 2);
        hashTable.put("orange", 3);
        
        System.out.println(hashTable.get("apple")); // 1
        System.out.println(hashTable.containsKey("banana")); // true
        
        hashTable.remove("orange");
    }
}

#Куча Фибоначи
import java.util.*;

public class FibonacciHeap {
    class Node {
        int key;
        int degree;
        boolean marked;
        Node parent, child, left, right;
        
        Node(int key) {
            this.key = key;
            this.degree = 0;
            this.marked = false;
            this.left = this;
            this.right = this;
        }
    }
    
    private Node minNode;
    private int count;
    
    public Node insert(int key) {
        Node newNode = new Node(key);
        if (minNode == null) {
            minNode = newNode;
        } else {
            addToRootList(newNode);
            if (key < minNode.key) {
                minNode = newNode;
            }
        }
        count++;
        return newNode;
    }
    
    private void addToRootList(Node node) {
        // Добавление в корневой список
    }
}
